Models/MuZero_agent_2.py

```diff
@@ -584,88 +589,136 @@ def player_to_play(player_num, action):
-                self.batch_expand_node(node[i], node[i].to_play, network_output)  # 7s
-                
-                # print("value {}".format(network_output["value"]))
-                self.backpropagate(search_path[i], network_output["value"].numpy()[i], min_max_stats[i], i)  # 12.08s
-                
-    def batch_policy(self, observation, prev_action):
-        self.NUM_ALIVE = observation.shape[0]
-        root = [Node(0) for _ in range(self.NUM_ALIVE)]
-        network_output = self.network.initial_inference(observation)  # 2.1 seconds
-        
-        for i in range(self.NUM_ALIVE):
-            self.batch_expand_node(root[i], i, network_output)  # 0.39 seconds
 
-            self.add_exploration_noise(root[i])
-            
-        self.run_batch_mcts(root, prev_action)  # 24.3 s (3 seconds not in that)
+            last_action = np.asarray(last_action)
+            network_output = self.network.recurrent_inference(np.asarray(hidden_states), last_action)   
+            value_prefix_pool = np.array(network_output["value_logits"]).reshape(-1).tolist()
+            value_pool = np.array(network_output["value"]).reshape(-1).tolist()
+            policy_logits_pool = np.array(network_output["policy_logits"]).tolist()
+
+            # add nodes to the pool after each search 
+            hidden_states_nodes = network_output["hidden_state"]
+            hidden_state_pool.append(hidden_states_nodes)
+
+            # reset 0
+            # reset the hidden states in LSTM every horizon steps in search
+            # only need to predict the value prefix in a range (eg: s0 -> s5)
+            assert horizons > 0
+            reset_idx = (np.array(search_lens) % horizons == 0)
+            assert len(reset_idx) == num
+            is_reset_lst = reset_idx.astype(np.int32).tolist()
+            # reset_idx = arr of t/f
+            # reset_idx ==> is_reset_lst
+            # tree node.isreset = is_reset_list[node]
+            hidden_state_index_x += 1
+
+            # backpropagation along the search path to update the attributes
+            tree.batch_back_propagate(hidden_state_index_x, discount, value_prefix_pool, value_pool, policy_logits_pool,
+                                      min_max_stats_lst, results, is_reset_lst)
+    
+    def batch_policy(self, observation):
+        self.NUM_ALIVE = observation.shape[0] 
+        # Setup specialised roots datastruction, format: env_nums, action_space_size, num_simulations
+        roots_cpp = tree.Roots(self.NUM_ALIVE, config.ACTION_DIM, config.NUM_SIMULATIONS) 
+        network_output = self.network.initial_inference(observation)  
+
+        value_prefix_pool = np.array(network_output["value_logits"]).reshape(-1).tolist()
+        policy_logits_pool = np.array(network_output["policy_logits"]).tolist()
+
+        # prepare the nodes to feed them into batch_mcts
+        noises = [np.random.dirichlet([config.ROOT_DIRICHLET_ALPHA] * config.ACTION_DIM).astype(np.float32).tolist()
```

Comment:
I do not know what the Dirichlet distribution is. Can you explain that? (I'm not asking because I think it doesn't work. I'm asking so we can have someone who understands exactly what this is doing so we can make adjustments and improvements later on as needed)

Reply:
I dont know, but it is the exact same method as used in out own code under add_exploration_noise (line 465 under MCTSAgent), so this isn't anything different to what we used before other than a slightly different format.